plugin/pkg/auth/authorizer/node/graph.go

```diff
@@ -0,0 +1,258 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"sync"
+
+	"k8s.io/kubernetes/pkg/api"
+	pvutil "k8s.io/kubernetes/pkg/api/persistentvolume"
+	podutil "k8s.io/kubernetes/pkg/api/pod"
+	"k8s.io/kubernetes/third_party/forked/gonum/graph"
+	"k8s.io/kubernetes/third_party/forked/gonum/graph/simple"
+)
+
+// namedVertex implements graph.Node and remembers the type, namespace, and name of its related API object
+type namedVertex struct {
+	vertexType vertexType
+	name       string
+	namespace  string
+	id         int
+}
+
+func newNamedVertex(vertexType vertexType, namespace, name string, id int) *namedVertex {
+	return &namedVertex{
+		vertexType: vertexType,
+		name:       name,
+		namespace:  namespace,
+		id:         id,
+	}
+}
+func (n *namedVertex) ID() int {
+	return n.id
+}
+func (n *namedVertex) String() string {
+	if len(n.namespace) == 0 {
+		return vertexTypes[n.vertexType] + ":" + n.name
+	}
+	return vertexTypes[n.vertexType] + ":" + n.namespace + "/" + n.name
+}
+
+// destinationEdge is a graph edge that includes a denormalized reference to the destination node
```

Comment:
nit: Add a comment to the effect that the denormalized destination is only relevant when there is a single leaf vertex reachable from the source vertex.
nit: s/node/vertex/